test/contracts/BullaInstaPay.test.ts

```diff
@@ -0,0 +1,116 @@
+import chai, { expect } from "chai";
+import { deployContract, solidity } from "ethereum-waffle";
+import fc from 'fast-check';
+import { ethers } from "hardhat";
+import BullaInstantPaymentMock from "../../artifacts/contracts/BullaInstaPay.sol/BullaInstantPayment.json";
+import { BullaInstantPayment } from "../../typechain/BullaInstantPayment";
+import { declareSignerWithAddress } from "../test-utils";
+import * as secp from "@noble/secp256k1";
+import keccak256 from "keccak256";
+import { ERC20 } from "../../typechain/ERC20";
+import ERC20Mock from "../../artifacts/contracts/mocks/BullaToken.sol/BullaToken.json";
+import { BigNumberish } from "ethers";
+
+chai.use(solidity);
+
+const ethAddressArb: () => fc.Arbitrary<string> = () => {
+    return fc.uint8Array({minLength: 31, maxLength: 33}).filter(arr => {try {secp.getPublicKey(arr); return true} catch(_) { return false;} }).map(secp.getPublicKey).map(x => ethers.utils.getAddress(`0x${keccak256(Buffer.from(x)).toString('hex').slice(-40)}`)) as fc.Arbitrary<string>;
+}
+
+const nonZeroIntArb = () => fc.bigUint().filter(x => x !== BigInt(0));
+const biggerThan10000Arb = () => nonZeroIntArb().filter(x => x.toString(10).length > 23); 
+const smallerThan10000Arb = () => nonZeroIntArb().filter(x => x.toString(10).length <= 23); 
+const nullAddress = "0x0000000000000000000000000000000000000000";
+
+describe("eth address arb", function () {
+    it("generates valid addresses", async function () {
+       fc.assert(fc.property(ethAddressArb(), address => {
+          return ethers.utils.isAddress(address.toLowerCase()) }))
+    });
+})
+
+describe("Bulla instant payment", function () {
+    let [signer, empty] = declareSignerWithAddress();
+    let bullaInstantPaymentContract: BullaInstantPayment;
+    let erc20Contract: ERC20;
+
+    this.beforeEach(async function () {
+        [signer, empty] = await ethers.getSigners();
+        erc20Contract = (await deployContract(signer, ERC20Mock)) as ERC20;
+        bullaInstantPaymentContract = (await deployContract(signer, BullaInstantPaymentMock)) as BullaInstantPayment;
+    });
+
+    const encodeInstantPaymentsCall = (toAddress: string, amount: BigNumberish, tokenAddress: string, description: string, tags: string[], ipfsHash: string) => bullaInstantPaymentContract.interface.encodeFunctionData('instantPayment', [toAddress, amount, tokenAddress, description, tags, ipfsHash]);
+
+    describe("batch", function () {
+        it("generates N events when N payments is made", function () {
+            return fc.assert(fc.asyncProperty(ethAddressArb(), fc.string(), smallerThan10000Arb(), fc.string(), fc.array(fc.string()), ethAddressArb(), fc.string(), smallerThan10000Arb(), fc.string(), fc.array(fc.string()), async (toAddressNative, descriptionNative, amountBigIntNative, ipfsHashNative, tagsNative, toAddressErc, descriptionErc, amountBigIntErc, ipfsHashErc, tagsErc,) => {
```

Comment:
small note: these lines are pretty long, let's make sure we run formatting on these files